Micropayment channel example polish

ChrisChinchilla opened the pull request:
I think this was the reference PR - https://github.com/ethereum/solidity/pull/4212

I tidied up some language, changed some of the code examples that were using wrong/outdated namespaces, or changed them to use namespaces that the docs of the project use.

Commit:
Begin review of micropayments example

Language tidy, add correct method and package namespaces and make more consistent with each project docs

First changes from review

Further fixes after review

Fix line breaks

Revert code changes

Update

chriseth commented on docs/solidity-by-example.rst:
```diff
@@ -645,99 +645,71 @@ Safe Remote Purchase
-In this section we will learn how to build a simple implementation
-of a payment channel. It use cryptographics signatures to make
+In this section we will learn how to build an example implementation
+of a payment channel. It use cryptographic signatures to make
```
Typo: `uses`

chriseth commented on docs/solidity-by-example.rst:
```diff
@@ -645,99 +645,71 @@ Safe Remote Purchase
 
 Imagine Alice wants to send a quantity of Ether to Bob, i.e.
 Alice is the sender and the Bob is the recipient.
+.. TODO: What does the below mean?
```
it means she does not have to send an actual transaction to the blockchain - is that what you mean?

ChrisChinchilla replied:
@chriseth I didn't write this, so it's not what 'I' mean, I was trying to understand what the original writer meant, 

> Alice only needs to send cryptographically signed messages off-chain
(e.g. via email) to Bob and it is similar to writing checks.

Just sounds kind of strange, but if it makes sense, then.

chriseth commented on docs/solidity-by-example.rst:
```diff
@@ -645,99 +645,71 @@ Safe Remote Purchase
-in a unusual way, instead of calling a function herself
-to initiate a payment, she will let Bob
-do that, and therefore pay the transaction fee.
+Alice and Bob use signatures to authorise transactions, which are a feature of smart contracts.
```
Hm, verifying signatures is something that can be done in smart contracts on Ethereum, but I would not call it a feature of smart contracts.

ChrisChinchilla replied:
Clarified

chriseth commented on docs/solidity-by-example.rst:
```diff
@@ -645,99 +645,71 @@ Safe Remote Purchase
-In this tutorial, we will sign messages in the browser
-using ``web3.js`` and ``MetaMask``.
-In particular, we will use the standard way described in `EIP-762 <https://github.com/ethereum/EIPs/pull/712>`_,
+Alice does not need to interact with the Ethereum network to sign the transaction, the process is completely offline.
```
Please try to change the newlines as little as possible (at least in such a large PR) as it makes the review much more complicated.

ChrisChinchilla replied:
@chriseth I always try, but when you rewrite and reorder entire sentences it's not always possible, and kind of makes looking and writing text challenging. When I review text I tend to review the whole thing, not diffs, as it's not quite the same as code.

I can reset all line breaks to a certain character number on every PR to keep it consistent, but no guarantee others will do that, and I don't know if that solves it anyway. Maybe this is something to discuss at some point.

Also in this PR submission there were lots of extraneous line breaks to remove.

chriseth commented on docs/solidity-by-example.rst:
```diff
@@ -645,99 +645,71 @@ Safe Remote Purchase
+    /// Hashing first makes things easier
+    var hash = web3.utils.sha3("message to sign");
+    web3.eth.personal.sign(hash, web3.eth.defaultAccount, function () {
+      console.log("Signed")
```
Perhaps this should better also output the signature.

ChrisChinchilla replied:
@chriseth In the original there was nothing inside the function and I added this mostly to know if it had worked. As the example adds to it later anyway, I'm thinking maybe just leave the function empty for now?

chriseth commented on docs/solidity-by-example.rst:
```diff
@@ -645,99 +645,71 @@ Safe Remote Purchase
-themselves, a so-called nonce, which is the number of transactions sent by an 
-account.
-The smart contract will check if a nonce is used multiple times.
+A replay attack is when a signed message is reused to claim authorisation for a second action. To avoid replay attacks we will use the same technique as in Ethereum transactions themselves, 'nonce', which is the number of transactions sent by an account. The smart contract will check if a nonce is used multiple times.
```
`a so-called "nonce"` sounds better to me.

ChrisChinchilla replied:
@chriseth not too keen on that phrase, try this as an alternative.

chriseth commented on docs/solidity-by-example.rst:
```diff
@@ -746,46 +718,28 @@ creates the proper signature for the ``ReceiverPays`` example:
-`ecrecover <mathematical-and-cryptographic-functions>`_
-that accepts a message along with the ``r``, ``s`` and ``v`` parameters and
-returns the address that was used to sign the message.
+In general, ECDSA signatures consist of two parameters, ``r`` and ``s``. Signatures in Ethereum include a third parameter called ``v``, that you can use to recover which account's private key was used to sign the message, and the transaction's sender. Solidity provides a built-in function `ecrecover <mathematical-and-cryptographic-functions>`_ that accepts a message along with the ``r``, ``s`` an ...
```
This sounds a little like you would be able to recover the private key which is of course not possible.  Also the transaction's sender can be derived from the public key, it is not something you could also recover using v.

chriseth commented on docs/solidity-by-example.rst:
```diff
@@ -746,46 +718,28 @@ creates the proper signature for the ``ReceiverPays`` example:
-We will use `inline assembly <assembly>`_ to do the job
-in the ``splitSignature`` function (the third function in the full contract
-at the end of this chapter).
+Signatures produced by web3.js are the concatenation of ``r``, ``s`` and ``v``, so the first step is to split these parameters apart. You can do this on the client-side, but doing it inside the smart contract means only one need to send one signature parameter needs rather than three. Splitting apart a byte array into component parts is a messy, so we use `inline assembly <assembly>`_ to do the j ...
```
"means only one need" reads a little weird

chriseth replied:
typo: "a messy"

chriseth commented:
Please rebase.

ChrisChinchilla commented:
@chriseth squashed and rebased

axic commented on docs/solidity-by-example.rst:
```diff
@@ -1101,30 +1031,28 @@ The recipient should verify each message using the following process:
     // this mimics the prefixing behavior of the eth_sign JSON-RPC method.
     function prefixed(hash) {
-        return ethereumjs.ABI.soliditySHA3(
+        return abi.soliditySHA3(
```
Why change this? The initial version is how the browser builds of ethereumjs work.

ChrisChinchilla replied:
@axic Ah, because when I looked at the project docs here - https://github.com/ethereumjs/ethereumjs-abi that's how they showed it and I don't see any other examples showing the original syntax. I was aiming for consistency among the projects to reduce confusion. Unless I'm missing something here (which is highly possible), I can't find any reference to the original syntax. This also could just mean that the original syntax should also be documented better.

axic replied:
The original syntax is when using the browser builds from https://github.com/ethereumjs/browser-builds

axic commented on docs/solidity-by-example.rst:
```diff
@@ -1101,30 +1031,28 @@ The recipient should verify each message using the following process:
-        var split = ethereumjs.Util.fromRpcSig(signature);
-        var publicKey = ethereumjs.Util.ecrecover(message, split.v, split.r, split.s);
-        var signer = ethereumjs.Util.pubToAddress(publicKey).toString("hex");
+        var split = ethereumjsutil.fromRpcSig(signature);
```
Why change this? The initial version is how the browser builds of ethereumjs work.

ChrisChinchilla replied:
@axic Pretty much the same as the comment above.

axic commented on docs/solidity-by-example.rst:
```diff
@@ -1079,16 +1012,13 @@ The full contract
-
-  
+.. note::
+  The function ``splitSignature`` is very simple and does not use all security checks. A real implementation should use a more rigorously tested library, such as openzepplin's `version <https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/ECRecovery.sol>`_ of this code.
```
Why change it from manually broken lines?

There are two problems with this:
- diffs are harder to see
- it is painful in some editors

ChrisChinchilla replied:
@axic The whole line break issue is a big conversation between writers and coders, but in all honestly it's mostly because I haven't worked on a code base for a while that hard enforced line breaks in docs and had just got out of the habit, fixed.

axic commented on docs/solidity-by-example.rst:
```diff
@@ -645,99 +645,71 @@ Safe Remote Purchase
-    web3.personal.sign(hash, web3.eth.defaultAccount, function () {...});
+    /// Hashing first makes things easier
+    var hash = web3.utils.sha3("message to sign");
+    web3.eth.personal.sign(hash, web3.eth.defaultAccount, function () {
```
I believe this is web3 1.x syntax. Should mention that above.

ChrisChinchilla commented:
@axic Ahh, I was using Node (which made more sense to me). OK, I'll check the code in a browser example as I'd like to understand how it all works anyway.

ChrisChinchilla commented:
OK @axic I reverted the code, I think I'll revisit this page soone though and make some of the examples clearer, or to be more precise, make it clearer how to write and run them.

ChrisChinchilla commented:
OK @axic I reverted the code, I think I'll revisit this page soone though and make some of the examples clearer, or to be more precise, make it clearer how to write and run them.

axic commented:
Please split the web3.js change into a separate PR. You are moving from the web3.js 0.x API to the 1.x API, and I believe the 0.x is still more common, given 1.x is not even released yet.

ChrisChinchilla commented:
@axic I reverted to the original text submitted by the original poster, so I'm not sure what the web3.js change you're referring to is? Unless I missed something somewhere or misunderstand you?

ChrisChinchilla commented:
Rebased and squashed. @chriseth I missed your comment (https://github.com/ethereum/solidity/pull/4523#discussion_r203402750) and attempted to address it, but after some searching, I wasn't necessarily any clearer, and it seems that there are other issues to document this better anyway (https://github.com/ethereum/solidity/issues/681). So I may address that separately when I look at that issue.

@axic still waiting for clarity on your comment (https://github.com/ethereum/solidity/pull/4523#issuecomment-423129476)

chriseth reviewed (approved):
I'm still not sure about the use of web3.js, but it's certainly better than before.
